File: leet_BTREE_1379_find_in_cloned___BFS.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function getTargetCopy(original: TreeNode | null, cloned: TreeNode | null, target: TreeNode | null): TreeNode | null {
    if (!original)
        return null
    if (original == target)
        return cloned
    let qo = [], qc = []
    qo.push(original)
    qc.push(cloned)
    while (qo.length != 0) {
        original = qo.shift()
        cloned = qc.shift()
        if (original == target)
            return cloned
        if (original.left)
            qo.push(original.left)
        if (original.right)
            qo.push(original.right)
        if (cloned.left)
            qc.push(cloned.left)
        if (cloned.right)
            qc.push(cloned.right)
    }
    return null
};

File: leet_BTREE_1379_find_in_cloned___BFS.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    TreeNode* getTargetCopy(TreeNode* o, TreeNode* c, TreeNode* t)
    {
        if (!o)
            return NULL;
        if (o == t)
            return c;

        deque<TreeNode*>    ori;
        deque<TreeNode*>    clo;

        ori.push_back(o);
        clo.push_back(c);
        while (!ori.empty())
        {
            o = ori.front();
            ori.pop_front();
            c = clo.front();
            clo.pop_front();
            if (o == t)
                return c;
            if (o->left)
                ori.push_back(o->left);
            if (o->right)
                ori.push_back(o->right);
            if (c->left)
                clo.push_back(c->left);
            if (c->right)
                clo.push_back(c->right);
        }
        return NULL;
    }
};
